kafka/socks5_wrapper.py

```diff
@@ -0,0 +1,236 @@
+                    self._sock.close()
+                    return errno.ECONNREFUSED
+
+                if buf[1:2] == b"\x00":
+                    # No authentication required
+                    self._state = ProxyConnectionStates.REQUEST_SUBMIT
+                elif buf[1:2] == b"\x02":
+                    userlen = len(self._proxy_url.username)
+                    passlen = len(self._proxy_url.password)
+                    self._buffer_out = struct.pack(
+                        "!bb{}sb{}s".format(userlen, passlen),
+                        1,
+                        userlen,
+                        self._proxy_url.username.encode(),
+                        passlen,
+                        self._proxy_url.password.encode(),
+                    )
+                    self._state = ProxyConnectionStates.AUTHENTICATING
+
+        if self._state == ProxyConnectionStates.AUTHENTICATING:
+            if self._flush_buf() is True:
+                buf = self._read_buf(2)
+                if buf[0:2] == b"\x01\x00":
+                    self._state = ProxyConnectionStates.REQUEST_SUBMIT
+                else:
+                    log.error("Socks5 proxy authentication failure")
+                    self._state = ProxyConnectionStates.DISCONNECTED
+                    self._sock.close()
+                    return errno.ECONNREFUSED
+
+        if self._state == ProxyConnectionStates.REQUEST_SUBMIT:
+            if self._target_afi == socket.AF_INET:
+                addr_type = 1
+                addr_len = 4
+            elif self._target_afi == socket.AF_INET6:
+                addr_type = 4
+                addr_len = 16
+            else:
+                log.error("Unknown address family, %r", self._target_afi)
+                self._state = ProxyConnectionStates.DISCONNECTED
+                self._sock.close()
+                return errno.ECONNREFUSED
+            self._buffer_out = struct.pack(
+                "!bbbb{}sh".format(addr_len),
+                5,  # version
+                1,  # command: connect
+                0,  # reserved
+                addr_type,
+                socket.inet_pton(self._target_afi, addr[0]),
+                addr[1],
+            )
+            self._state = ProxyConnectionStates.REQUESTING
+
+        if self._state == ProxyConnectionStates.REQUESTING:
+            if self._flush_buf() is True:
+                buf = self._read_buf(2)
+                if buf[0:2] == b"\x05\x00":
+                    self._state = ProxyConnectionStates.READ_ADDRESS
+                else:
+                    log.error("Proxy request failed: %r", buf[1:2])
+                    self._state = ProxyConnectionStates.DISCONNECTED
+                    self._sock.close()
+                    return errno.ECONNREFUSED
+
+        if self._state == ProxyConnectionStates.READ_ADDRESS:
+            # we don't really care about the remote endpoint address, but need to clear the stream
+            buf = self._peek_buf(2)
```

Comment:
Any reason why peek is used instead of read here? Looks like we'll anyway want to consume the bytes.

Reply:
Peek here allows us to return to the same state machine step if we would capture expected address type but not yet have all data available for the entire address. It's just to avoid an extra state machine step and/or maintaining additional local state on the already received bytes.